### PR TITLE
Support `torch.linalg.trace`

### LINKED ISSUE #62255: Support `torch.linalg.trace`
## 🚀 Feature
Support `torch.linalg.trace`.
In addition to the features of `torch.trace`, `axis1`, `axis2` and `offset` options are also wanted.
(cc: @mruberry, @rgommers, @emcastillo and @kmaehashi)

## Motivation and Pitch
To enhance the compatibility with NumPy’s interface and Python array API standard.

## Needs discussions
Should the axis name be `axis1` and `axis2`, or `dim0` and `dim1`? Both should be supported?

## References
* [Python array API linalg.trace](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#linalg-trace-x-axis1-0-axis2-1-offset-0)

### PR DESCRIPTION
**Pull request by asi1024**
Fixes #62255 (cc/ @mruberry, @rgommers, @emcastillo, @kmaehashi)

This PR adds support of `torch.linalg.trace` for the compatibility with NumPy's interface and Python array API standard.

```py
>>> torch.linalg.trace(torch.arange(18).reshape(2, 3, 3))
tensor([12, 14, 16])
```

TODO:
 - [x] Add documentation
 - [x] Add tests

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

**Comment by mruberry** [comment]
Hey @asi1024, just checking in on this PR because it's still marked as "draft." Is it ready for a review?

**Comment by asi1024** [comment]
@mruberry Sorry for my late response. I will mark "ready for review" after adding tests and documentation!

File: aten/src/ATen/native/native_functions.yaml
Comment (IvanYashchuk) [inline aten/src/ATen/native/native_functions.yaml:6224]: This line is overwritten by `CompositeExplicitAutograd: linalg_trace`. The code in the `linalg_trace` function is independent of the device so `CPU, CUDA` specialization is not needed here and `CompositeExplicitAutograd` is the correct choice of the dispatch key.
```suggestion

```

File: aten/src/ATen/native/native_functions.yaml
Comment (IvanYashchuk) [inline aten/src/ATen/native/native_functions.yaml:6227]: Could you please remove this entry from native_functions.yaml?
Most of the backward functions in PyTorch are placed in `torch\csrc\autograd\FunctionsManual.cpp` and `torch\csrc\autograd\FunctionsManual.h`, so let's move `linalg_trace_backward` from `ReduceOps.cpp`.

File: aten/src/ATen/native/ReduceOps.cpp
Comment (IvanYashchuk) [inline aten/src/ATen/native/ReduceOps.cpp:1128]: Is this change needed in this PR?
The function `prod(Tensor self, int dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor` should be autogenerated with https://github.com/pytorch/pytorch/blob/54d060a8c91c27167925b380698331850352c841/aten/src/ATen/native/native_functions.yaml#L4200

File: aten/src/ATen/native/native_functions.yaml
Comment (IvanYashchuk) [reply aten/src/ATen/native/native_functions.yaml:6224]: On a second thought using `CompositeImplicitAutograd` should be better, then the backward function is not needed.

**Review by IvanYashchuk — COMMENTED** [review]
Hey, @asi1024, thanks for your contribution! I left a suggestion to use the `CompositeImplicitAutograd` dispatch key that would allow us to remove the `_backward` function trimming down unnecessary code. After that, I think the PR should be good to go.

File: torch/linalg/__init__.py
Comment (Lezcano) [inline torch/linalg/__init__.py:2049]: What diagonal? Given that we have the parameter `offset`, this should probably read:

>Returns the sum of the elements of a diagonal.

Followed by an explanation of how the `offset` parameter chooses a diagonal.

File: torch/linalg/__init__.py
Comment (Lezcano) [inline torch/linalg/__init__.py:2050]: It is missing a description of the arguments.

File: test/test_linalg.py
Comment (Lezcano) [inline test/test_linalg.py:8019]: The call to `torch.from_numpy` here and a few lines below is not necessary as `assertEqual` is able to compare tensors and numpy arrays. Even more, not calling it is often faster. Same below.

File: test/test_linalg.py
Comment (Lezcano) [inline test/test_linalg.py:8018]: This might work without all the explicit castings by simply doing

```suggestion
            yn = np.trace(x.cpu(), axis1=-2, axis2=-1)
```
Same below.

File: test/test_linalg.py
Comment (Lezcano) [inline test/test_linalg.py:8020]: Given that the offset zero is already tested below, there is no need to test all this here. You can just do just inside the offsets. Even better this function could simply be something along the lines of:
```python
for shape, offset in product(shapes, offsets):
    size = np.prod(shape)
    x = torch.arange(size, device=device).reshape(*shape)
    y = torch.linalg.trace(x, offset)
    yn = np.trace(x.cpu(), offset, axis1=-2, axis2=-1)
    self.assertEqual(y, yn)
```

**Review by lezcano — COMMENTED** [review]
Left a few points regarding docs / testing.

File: test/test_linalg.py
Comment (asi1024) [reply test/test_linalg.py:8020]: L8017-L8020 is a test to check if the default value of `offset` is 0.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:4202]: As @IvanYashchuk pointed out, this is not necessary. Just remove the lines 
```
  dispatch:
    CompositeExplicitAutograd: linalg_trace
```
from `native_functions.yaml` and autodiff will be able to generate an efficient derivative.

File: torch/linalg/__init__.py
Comment (Lezcano) [inline torch/linalg/__init__.py:2049]: Nit: returned

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:7340]: Once you remove the explicit derivative, you should be able to add the line
```python
supports_forward_ad=True,
```

**Review by lezcano — COMMENTED** [review]
Just a few things and this should be ready to go :)

File: torch/linalg/__init__.py
Comment (Lezcano) [inline torch/linalg/__init__.py:2049]: Nit. returned* :D

File: torch/csrc/autograd/FunctionsManual.h
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.h:304]: This was left hanging.

**Review by lezcano — COMMENTED** [review]
Just two small things and this should be ready.  cc @mruberry

**Review by lezcano — APPROVED** [review]
LGTM!

**Comment by asi1024** [comment]
@Lezcano @IvanYashchuk Thanks for your review! :smile:

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:1082]: File a new issue to improve the performance of linalg.trace and point to that issue in this comment

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:6238]: Let's make offset kwarg-only to be consistent with the Python Array API

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2047]: input, *, offset=0

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2049]: "Computes the trace of a matrix or a batch of traces for a batch of matrices. The trace of a matrix is the sum of elements along its diagonal..."

Then this needs to explain what the diagonal is and how offset works. A proper mathematical formula would help. An issue with the current description is that `i` is not well defined. We may also want to link the Wikipedia article, which does a good job explaining what the diagonal is (https://en.wikipedia.org/wiki/Main_diagonal).

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2054]: kwarg

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2055]: Does this wrap?

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2053]: matrix (or batch of matrices) to compute the trace (or traces) of.

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2059]: When working with a single tensor prefer `t` for the name, when working with multiple prefer a, b, c... when the tensors don't already have natural names.

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2061]: This is cool! Add an example showing how to use offset, too

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2010]: We should probably shrink these sample inputs -- they seem very big and that will make the tests take longer than necessary

**Review by mruberry — COMMENTED** [review]
Hey @asi1024! This is really cool. Remember to update the .rst file, too, so this appears in the docs.

I made a few comments inline for your review. In addition I suggest filing a follow-up issue, but I think now it's better to address that issue in this PR. In particular, I think we should resolve the performance issue by keeping the cpu-specific optimization implementing by torch.trace today, and we should make torch.trace an alias of torch.linalg.trace by updating it take an offset kwarg too, and having its implementation defer to torch.linalg.trace. The OpInfos for the two can then be combined, too.

Looking forward to hearing your thoughts!

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [reply aten/src/ATen/native/ReduceOps.cpp:1082]: Actually (see below) I recommend we address this in this PR and not file a follow-up issue.

**Comment by mruberry** [comment]
Hey @asi1024! Catching up on reviews now and wanted to check-in on this PR. How's it going?

**Comment by Lezcano** [comment]
@asi1024 what's the status of this PR?

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [inline aten/src/ATen/native/ReduceOps.cpp:1084]: This is also missing a `TORCH_CHECK(self.dim() >= 2, ...`

File: torch/linalg/__init__.py
Comment (Lezcano) [reply torch/linalg/__init__.py:2049]: To keep the linalg docs with a homogeneous notation, I'm proposing writing something like:

> trace(input, *, offset=0) -> Tensor
> Computes the trace of a square matrix.
>
> Letting :math:`\mathbb{K}` be :math:`\mathbb{R}` or :math:`\mathbb{C}`,
> the **trace ** of a square matrix
> :math:`A \in \mathbb{K}^{n \times n}` is defined as
>
> .. math::
>
>     \operatorname{tr}(A) = \sum_{i=0}^{n-1} A_{i,i}
>
> If :attr:`offset` is provided, the sum will run over the elements :math:`A_{i, j}`
> with :math:`i - j = \texttt{offset}`.
>
> Supports input of float, double, cfloat and cdouble dtypes.
> Also supports batches of matrices, and if :attr:`A` is a batch of matrices then
> the output has the same batch dimensions.
>
> Args:
>     A (Tensor): tensor of shape `(*, n, n)` where `*` is zero or more batch dimensions.
>
> Keyword args:
>     offset (int, optional): offset of the diagonal from the main diagonal. A positive value indicates
>                             an upper diagonal, and negative value a lower diagonal. Default: `0`.

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:7339]: This line may not be necessary (not 100% sure)

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:7336]: Now that we are at it, it would be good to make this operation support out, as it would take very little effort to do so.

Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py:7336]: Do you think should `out` be a keyword only argument or not?
Other PyTorch functions can take `out` as a positional argument, but `torch.linarg.trace(A, out, *, offset)` is confusing because it is not compatible with `numpy.trace(A, offset)`.

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:7336]: `out` is always kwonly and the last output in PyTorch, as such it should be `torch.linarg.trace(A, *, offset=0, out=None)`

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:1084]: I think it'd also be nice to check that offset is in the range [-dim, dim) and throw a nice error message otherwise.

File: torch/testing/_internal/common_methods_invocations.py
Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py:7339]: `torch.linalg.trace` allows non-square matrix as an input.
How about saying that the behavior is equivalent to `torch.diagonal(A, offset, -2, -1).sum(-1)`?

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:7339]: This is an extremely odd thing. The trace of a matrix is only mathematically defined for square matrices. I'm opening a PR in the Array API about this because this does not seem right.

I think this comment was supposed to go under the docs comment (which is an answer to a comment above, that's why you cannot answer directly to it here).

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:7339]: Let's wait until https://github.com/data-apis/array-api/issues/359 is resolved for this one.

**Review by lezcano — COMMENTED** [review]
So, as far as I can see, there's just a couple details missing:
- Apply the suggested changes (or suggest some new ones) when it comes to the docs https://github.com/pytorch/pytorch/pull/62714#discussion_r777940303
- Throw a nice message when the offset is out of the range https://github.com/pytorch/pytorch/pull/62714#discussion_r778831325
- Make the inputs for the opinfo smaller https://github.com/pytorch/pytorch/pull/62714#discussion_r720122625

**Comment by Lezcano** [comment]
On a side note, about this function, @mruberry have a look at https://github.com/data-apis/array-api/issues/359. Not a hill to die on, but I still do not understand the decision there.

**Comment by mruberry** [comment]
> On a side note, about this function, @mruberry have a look at [data-apis/array-api#359](https://github.com/data-apis/array-api/issues/359). Not a hill to die on, but I still do not understand the decision there.

That is interesting but we might as well be consistent.

**Comment by Lezcano** [comment]
Any update on this one @asi1024 ?

**Comment by asi1024** [comment]
@Lezcano Thanks for your reviewing! :smile:

> So, as far as I can see, there's just a couple details missing:
> 
> * Apply the suggested changes (or suggest some new ones) when it comes to the docs [#62714 (comment)](https://github.com/pytorch/pytorch/pull/62714#discussion_r777940303)

In https://github.com/data-apis/array-api/issues/359, it was decided to support non-square matrices in `linalg.trace`. Could you please guide me if I still need to modify the documentation? :smiley: 

> * Throw a nice message when the offset is out of the range [#62714 (comment)](https://github.com/pytorch/pytorch/pull/62714#discussion_r778831325)

Hmm, `numpy.trace` returns `0` if `offset` is out of the range.

> * Make the inputs for the opinfo smaller [#62714 (comment)](https://github.com/pytorch/pytorch/pull/62714#discussion_r720122625)

I have shrunk the sample inputs in cecc545c, but do I need to shrink them more?

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:9995]: Doesn't it? I don't know what happened with this discussion https://github.com/pytorch/pytorch/pull/62714#discussion_r778830799

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:9992]: It'd be good to add a line to the OpInfo:
```python
ref=np.trace,
```
this way we would check that the implementation is correct, comparing it against numpy.

File: torch/testing/_internal/common_methods_invocations.py
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:2581]: This one should probably go.

**Review by lezcano — COMMENTED** [review]
Re the previous comment, just one thing, everything else is fine.
We do want to error out if the offset is too big or too small (I discussed this with @mruberry and he agreed). If people complain, we can always revert.

Similar issue: https://github.com/pytorch/pytorch/issues/71204

File: torch/testing/_internal/common_methods_invocations.py
Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py:9995]: The tests seem to fail if enabling supports_out. Perhaps this option is assuming that `out` is a positional argument?

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:9995]: What's the error it throws? `out=` should be a kwonly argument, so it's not clear to me a priori what's the problem there...

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [inline aten/src/ATen/native/ReduceOps.cpp:1128]: This should be `auto k = std::min(self.size(-2), self.size(-1));` and then use `k` in the check, right?

Comment (asi1024) [reply aten/src/ATen/native/ReduceOps.cpp]: I fixed the condition to `-self.size(-2) < offset && offset < self.size(-1)`. linalg.trace should accept `offset=9` for an input of size (1, 10), for example.

File: torch/testing/_internal/common_methods_invocations.py
Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py]: Ah, sorry! The test failure was due to the missing `check_scalar_type_device_layout_equal(result, self);`!

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp]: You are completely right, sorry for that.

**Review by lezcano — APPROVED** [review]
I think that this comment: https://github.com/pytorch/pytorch/pull/62714/files#r777940303 slipped through the cracks.

That's the only thing though. Otherwise this PR lgtm provided the CI agrees. cc @mruberry

**Comment by asi1024** [comment]
@Lezcano Now all CIs have passed! PTAL!

**Comment by Lezcano** [comment]
Yeah, this looks good to me. I just found (yet another, ugh, sorry).
We're missing to add the relevant entry to the docs in `docs/source/linalg.rst`! It could probably go between `diagonal` and `det`.

Sorry for that! I believe this is the last missing thing! :D
Otherwise, we can just wait for @mruberry to review.

**Comment by asi1024** [comment]
@Lezcano The CI failures look unrelated to this PR. Could you take another look?

File: torch/linalg/__init__.py
Comment (mruberry) [inline torch/linalg/__init__.py:2385]: This is really well written.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2581]: Add a sample with a negative offset and a comment explaining the format of these tuples

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9997]: Nice reference

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:7767]: Nice test

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2579]: What happens on empty tensors? We should probably add a case for them. What about a batched sample input too? `(S, S, S`

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:7770]: What happens if offset is an absurd number, like 100?

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:7769]: Adding the empty case here would be interesting, too

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:6506]: Nice schemas

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:1133]: The user documented name is `input` (per your docs below) and these warnings should start with the name of the operation like this:

```
torch.linalg.trace(): input should have at least...
```

**Review by mruberry — COMMENTED** [review]
Hey @asi1024! Overall this looks really good but there are a few comments inline from me and @Lezcano that still need to be addressed. Just ping me when they're done and we'll get this merged.

File: aten/src/ATen/native/ReduceOps.cpp
Comment (Lezcano) [reply aten/src/ATen/native/ReduceOps.cpp:1133]: It might be nice to change the user-facing name of this argument to `A`, which is the name we use throughout `torch.linalg`

File: test/test_linalg.py
Comment (asi1024) [reply test/test_linalg.py:7770]: `RuntimeError` will be raised if `offset` is out of range. I will add a test for this case!

Comment (mruberry) [reply test/test_linalg.py:7770]: > `RuntimeError` will be raised if `offset` is out of range. I will add a test for this case!

Make it an ErrorInput

File: torch/testing/_internal/common_methods_invocations.py
Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py:2579]: The trace function implemented in this PR returns different values from `numpy.trace` for 3-dim inputs. `numpy.trace` reduces with axis1=0, axis2=1 whereas array API specifies to reduce with axis1=-2, axis2=-1.

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:2579]: Is it possible to compare then against a lambda with our same defaults that the  calls into np.trace?

**Pull request closed without merging**
